### PR TITLE
Write batch scripts for multi-line commands with CMD.EXE

### PR DESCRIPTION
**Pull request by lox**
In #648 we allowed for the default shell to be configured and we stopped writing temporary scripts for commands to be executed with in favour of directly invoking them with the configured shell command. This broke multi-line commands for CMD.EXE, as it seems to ignore everything after the first newline.

The future is likely powershell for Windows users, but this writes batch files if a CMD.EXE shell is detected to maintain backwards compatibility.

This also introduces a new library I wrote for parsing and quoting strings with specific shell semantics: buildkite/shellwords. This is used to provide a more robust solution to #683.

### CODE REVIEW

**Comment by lox** [comment]
This is ready for code-review now @keithpitt! ✈️

File: bootstrap/bootstrap.go [hunk 1]
```diff
@@ -890,38 +893,98 @@ func (b *Bootstrap) defaultCommandPhase() error {
...
 	}
 
-	var cmd []string
+	if len(shell) == 0 {
+		return fmt.Errorf("No shell set for bootstrap")
+	}
 
-	if commandIsScript {
+	// Windows CMD.EXE is horrible and can't handle newline delimited commands. We write
+	// a batch script so that it works, but we don't like it
```
Comment (sj26) [inline bootstrap/bootstrap.go]: :disappointed:

File: bootstrap/bootstrap.go [hunk 2]
```diff
@@ -890,38 +893,98 @@ func (b *Bootstrap) defaultCommandPhase() error {
...
+
+	// The shell gets parsed based on the operating system
+	shell, err := shellwords.Split(b.Shell)
 	if err != nil {
-		return err
+		return fmt.Errorf("Failed to split shell (%q) into tokens: %v", b.Shell, err)
 	}
 
-	var cmd []string
+	if len(shell) == 0 {
```
Comment (keithpitt) [inline bootstrap/bootstrap.go]: What sort of shell values can be passed to cause it to get a `len` of 0?

File: clicommand/bootstrap.go [hunk 3]
```diff
@@ -222,6 +222,7 @@ var BootstrapCommand = cli.Command{
 			Name:   "shell",
 			Usage:  "The shell to use to interpret build commands",
```
Comment (keithpitt) [inline clicommand/bootstrap.go]: Should we mention "This defaults to our best guess at what makes sense based on the current operation system" or something? Just thinking about what we copy/paste into the docs for this `--help` output.

Comment (lox) [reply clicommand/bootstrap.go]: Yeah, makes sense.

File: bootstrap/bootstrap.go (hunk 2 above)
Comment (lox) [reply bootstrap/bootstrap.go]: An empty string :)

File: clicommand/bootstrap.go (hunk 3 above)
Comment (lox) [reply clicommand/bootstrap.go]: Oh, this is the bootstrap cli command. Did you mean here or the agent start command?

Comment (keithpitt) [reply clicommand/bootstrap.go]: > Did you mean here or the agent start command?

Oh yeah, I meant that!

Comment (lox) [reply clicommand/bootstrap.go]: "The shell to use to interpret build commands. Defaults to bash or cmd.exe depending on your operating system."

Comment (keithpitt) [reply clicommand/bootstrap.go]: 👍

Comment (lox) [reply clicommand/bootstrap.go]: Or maybe "The shell commands used to interpret build commands, e.g /bin/bash -e -c"

Comment (keithpitt) [reply clicommand/bootstrap.go]: Yeah that works as well

**Pull request merged**